### PR TITLE
Add LoggingLevelSwitch.MinimumLevelChanged

### LINKED ISSUE #1468: Allow developers to react to changes in the MinimumLevel of a LoggingLevelSwitch
**Is your feature request related to a problem? Please describe.**
There's currently no easy way of detecting when the `MinimumLevel` of a `LoggingLevelSwitch` changes, and the recommended workaround is monitoring its value on a timer, which works but is cumbersome.

**Describe the solution you'd like**

There has been requests to make the `MinimumLevel` virtual so it can be overridden, or change it to an interface for custom implementations, (e.g. #1436, #1417, #1232,  and #682), and none of these were implemented because it would considerably affect performance of reads on the `MinimumLevel`.

These requests seem to share the same end goal: Run custom code **when the `MinimumLevel` changes**.

Given that _writes_ to `MinimumLevel` shouldn't happen nearly as often as _reads_, it seems that executing custom code immediately after the value of `MinimumLevel` is changed wouldn't be a big concern in terms of performance - unless I'm missing something (?).

With that in mind, I'd like to propose the introduction of a couple of methods in the `LoggingLevelSwitch` class that allow developers to enable/disable running custom code when the `MinimumLevel` changes.

It could be something along these lines:

```diff
public class LoggingLevelSwitch
{
+   Action<LogEventLevel> _notify;
    // ... (omitted for brevity)

    public LogEventLevel MinimumLevel
    {
        get { return _minimumLevel; }
        set
        {
            _minimumLevel = value;
+           var n = _notify;
+           n?.Invoke(value);
        }
    }
+
+   public void EnableNotifications(Action<LogEventLevel> notify)
+   {
+      _notify = notify ?? throw new ArgumentNullException(nameof(notify));
+   }
+
+   public void DisableNotifications()
+   {
+       _notify = null;
+   }
}
```

**Describe alternatives you've considered**
Monitor instances of `LoggingLevelSwitch` on a timer or background thread.

**Additional context**

If executing custom code in the setter of `MinimumLevel` turns out not to be a bad idea, we could also consider:

- Checking if the new value being assigned is different than the current value and only run the action if that's the case
- Provide previous and actual `LogEventLevel` rather than only the actual
- Allow more than one `Action` to be executed rather than a single one w/ last one wins

I'd be happy to send a PR once a design is agreed.

### PR DESCRIPTION
**Pull request by sungam3r**
fixes #1468

### CODE REVIEW

File: src/Serilog/Core/LoggingLevelSwitch.cs [hunk 1]
```diff
@@ -20,6 +20,7 @@ namespace Serilog.Core;
 public class LoggingLevelSwitch
 {
     volatile LogEventLevel _minimumLevel;
+    readonly object locker = new();
```
Comment (nblumhardt) [inline src/Serilog/Core/LoggingLevelSwitch.cs:23]: I think this is `_sync` or a similar name in most of the codebase?

File: src/Serilog/Core/LoggingLevelSwitch.cs [hunk 2]
```diff
@@ -31,6 +32,11 @@ public LoggingLevelSwitch(LogEventLevel initialMinimumLevel = LogEventLevel.Info
         _minimumLevel = initialMinimumLevel;
     }
 
+    /// <summary>
+    /// The event arises when <see cref="MinimumLevel"/> changed.
```
Comment (nblumhardt) [inline src/Serilog/Core/LoggingLevelSwitch.cs:36]: Needs a comment pointing out that the event is raised under a lock (which will lead to deadlock if a handler waits on another lock `B`, which is already held by a thread that's attempting to change the level and is blocked on our lock here).

File: src/Serilog/Core/LoggingLevelSwitch.cs [hunk 3]
```diff
@@ -40,6 +46,44 @@ public LoggingLevelSwitch(LogEventLevel initialMinimumLevel = LogEventLevel.Info
...
+                }
+            }
+        }
+    }
+}
+
+/// <summary>
+/// Event arguments for <see cref="LoggingLevelSwitch.MinimumLevelChanged"/> event.
+/// </summary>
+public class LoggingLevelSwitchChangedEventArgs : EventArgs
```
Comment (nblumhardt) [inline src/Serilog/Core/LoggingLevelSwitch.cs:67]: Needs its own file.

**Review by nblumhardt — COMMENTED** [review]
Had a quick scan through, it looks good to me, might be worth popping a comment on the original issue thread asking for feedback?

File: src/Serilog/Core/LoggingLevelSwitch.cs (hunk 1 above)
Comment (sungam3r) [reply src/Serilog/Core/LoggingLevelSwitch.cs:23]: изображение

File: src/Serilog/Core/LoggingLevelSwitch.cs (hunk 2 above)
Comment (sungam3r) [reply src/Serilog/Core/LoggingLevelSwitch.cs]: done

File: src/Serilog/Core/LoggingLevelSwitch.cs (hunk 3 above)
Comment (sungam3r) [reply src/Serilog/Core/LoggingLevelSwitch.cs]: done

File: src/Serilog/Core/LoggingLevelSwitch.cs (hunk 1 above)
Comment (nblumhardt) [reply src/Serilog/Core/LoggingLevelSwitch.cs:23]: Looks like in the shipping code the only one we have is `_templatesLock`. I think this suggests a name like `_levelUpdateLock` perhaps?

(No idea what the `lock(new object())` in that test is expecting to do :-))

Comment (sungam3r) [reply src/Serilog/Core/LoggingLevelSwitch.cs]: done

File: test/Serilog.Tests/Core/LoggerTests.cs [hunk 4]
```diff
@@ -137,7 +137,7 @@ public void TheNoneLoggerIsAConstant()
     [Fact]
     public void TheNoneLoggerIsSingleton()
     {
-        lock (new object())
+        lock (this)
```
Comment (nblumhardt) [inline test/Serilog.Tests/Core/LoggerTests.cs:140]: I don't think any other code locks `this` so there's probably still no synchronization going on here; I'll merge this change as I think it shows the intent of the old code better, but we'll need to configure xUnit to serialize test execution around the static `Log` class separately, if there's any code outside of `LoggerTests` that interacts with it.

**Review by nblumhardt — APPROVED** [review]

**Pull request merged**